src/toxencryptsave/encryptsave.rs

```diff
@@ -0,0 +1,165 @@
+/*
+    Copyright © 2016 quininer kel <quininer@live.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+use sodiumoxide::crypto::pwhash::{
+    SALTBYTES, MEMLIMIT_INTERACTIVE, OPSLIMIT_INTERACTIVE,
+    Salt, OpsLimit,
+    gen_salt, derive_key
+};
+use sodiumoxide::crypto::box_::{
+    PRECOMPUTEDKEYBYTES, NONCEBYTES, MACBYTES,
+    Nonce, PrecomputedKey,
+    gen_nonce
+};
+use sodiumoxide::crypto::hash::sha256;
+use ::toxcore::crypto_core;
+
+pub const MAGIC_LENGTH: usize = 8;
+pub const MAGIC_NUMBER: &'static [u8; MAGIC_LENGTH] = b"toxEsave";
+pub const SALT_LENGTH: usize = SALTBYTES;
+pub const KEY_LENGTH: usize = PRECOMPUTEDKEYBYTES;
+pub const EXTRA_LENGTH: usize = MAGIC_LENGTH + SALT_LENGTH + NONCEBYTES + MACBYTES;
+
+
+#[derive(Clone, Debug)]
+pub struct PassKey {
+    pub salt: Salt,
+    pub key: PrecomputedKey
+}
+
+impl PassKey {
+    pub fn new(passphrase: &[u8]) -> Result<PassKey, KeyDerivationError> {
+        PassKey::with_salt(passphrase, gen_salt())
+    }
+
+    pub fn with_salt(passphrase: &[u8], salt: Salt) -> Result<PassKey, KeyDerivationError> {
+        if passphrase.is_empty() { return Err(KeyDerivationError::Null) };
+
+        let OpsLimit(ops) = OPSLIMIT_INTERACTIVE;
+        let sha256::Digest(passhash) = sha256::hash(passphrase);
+        let maybe_key = PrecomputedKey::from_slice(try!(
+            derive_key(
+                &mut [0; KEY_LENGTH],
+                &passhash,
+                &salt,
+                OpsLimit(ops * 2),
+                MEMLIMIT_INTERACTIVE
+            ).or(Err(KeyDerivationError::Failed))
+        ));
+
+        Ok(PassKey {
+            salt: salt,
+            key: try!(maybe_key.ok_or(KeyDerivationError::Failed))
+        })
+    }
+
+    pub fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, Encryption> {
```

Comment:
Can you help me to complete some documents?
My mother tongue is not English, not good at writing the document.

在 2016年05月20日 00:05, zetok 写道:
>
> In src/toxencryptsave/encryptsave.rs
> <https://github.com/zetok/tox/pull/9#discussion_r63906868>:
>
> > +            derive_key(
> > +                &mut [0; KEY_LENGTH],
> > +                &passhash,
> > +                &salt,
> > +                OpsLimit(ops * 2),
> > +                MEMLIMIT_INTERACTIVE
> > +            ).or(Err(KeyDerivationError::Failed))
> > +        ));
> > +
> > +        Ok(PassKey {
> > +            salt: salt,
> > +            key: try!(maybe_key.ok_or(KeyDerivationError::Failed))
> > +        })
> > +    }
> > +
> > +    pub fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>, Encryption> {
>
> Needs docs that will say when this fn can fail.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/zetok/tox/pull/9/files/e2a279ff3c6d2be9e0d2e5e167e5302cf9415b47#r63906868>
>